store/s3storeKMS.go

```diff
@@ -0,0 +1,486 @@
+package store
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/pkg/errors"
+)
+
+// latest is used to keep a single object in s3 with all of the
+// most recent values for the given service's secrets.  Keeping this
+// in a single s3 object allows us to use a single s3 GetObject
+// for ListRaw (and thus chamber exec).
+type LatestIndexFile struct {
+	Latest map[string]LatestValue `json:"latest"`
+}
+
+type LatestValue struct {
+	Version  int    `json:"version"`
+	Value    string `json:"value"`
+	KMSAlias string `json:"KMSAlias"`
+}
+
+var _ Store = &S3KMSStore{}
+
+type S3KMSStore struct {
+	svc    s3iface.S3API
+	stsSvc *sts.STS
+	bucket string
+}
+
+// Deprecated; use NewS3KMSStoreWithBucket instead
+func NewS3KMSStore(numRetries int) (*S3KMSStore, error) {
+	bucket, ok := os.LookupEnv(BucketEnvVar)
+	if !ok {
+		return nil, fmt.Errorf("Must set %s for s3 backend", BucketEnvVar)
+	}
+
+	return NewS3KMSStoreWithBucket(numRetries, bucket)
+}
+
+func (s *S3KMSStore) KMSKey() string {
+	fromEnv, ok := os.LookupEnv("CHAMBER_KMS_KEY_ALIAS")
+	if !ok {
+		return DefaultKeyID
+	}
+	if !strings.HasPrefix(fromEnv, "alias/") {
+		return fmt.Sprintf("alias/%s", fromEnv)
+	}
+
+	return fromEnv
+}
+
+func NewS3KMSStoreWithBucket(numRetries int, bucket string) (*S3KMSStore, error) {
+	session, region, err := getSession(numRetries)
+	if err != nil {
+		return nil, err
+	}
+
+	svc := s3.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	stsSvc := sts.New(session, &aws.Config{
+		MaxRetries: aws.Int(numRetries),
+		Region:     region,
+	})
+
+	return &S3KMSStore{
+		svc:    svc,
+		stsSvc: stsSvc,
+		bucket: bucket,
+	}, nil
+}
+
+func (s *S3KMSStore) Write(id SecretId, value string) error {
+	index, err := s.readLatest(id.Service)
+	if err != nil {
+		return err
+	}
+
+	if val, ok := index.Latest[id.Key]; ok {
+		if val.KMSAlias != s.KMSKey() {
```

Comment:
I wonder if this comparison is robust enough? You could have multiple aliases referring to the same key, right? Not sure what's desirable here.

Reply:
See https://github.com/segmentio/chamber/pull/195#discussion_r266699280